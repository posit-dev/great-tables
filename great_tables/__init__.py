# Set version ----
from importlib_metadata import version as _v

__version__ = _v("great_tables")

del _v

# Main gt imports ----

from .gt import GT
from . import vals, loc, style
from ._styles import FromColumn as from_column
from ._helpers import (
    letters,
    LETTERS,
    px,
    pct,
    md,
    html,
    google_font,
    random_id,
    system_fonts,
    define_units,
    nanoplot_options,
)


__all__ = (
    "GT",
    "exibble",
    "letters",
    "LETTERS",
    "px",
    "pct",
    "md",
    "html",
    "google_font",
    "system_fonts",
    "define_units",
    "nanoplot_options",
    "random_id",
    "from_column",
    "vals",
    "loc",
    "style",
)


def __getattr__(k: str):
    # exibble dataset available on top-level module, but is a pandas DataFrame.
    # Since pandas is an optional dependency, we import exibble dynamically.
    if k == "exibble":
        from great_tables.data import exibble

        return exibble

    # allow the data submodule to be accessed directly, as if it were a top-level import
    if k == "data":
        import great_tables.data

        return great_tables.data

    raise AttributeError(f"cannot get attribute {k} from great_tables ({__file__})")
