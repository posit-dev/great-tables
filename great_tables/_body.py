from __future__ import annotations

from typing import TYPE_CHECKING

from ._tbl_data import copy_data

if TYPE_CHECKING:
    from ._gt_data import Body


def body_reassemble(body: Body) -> Body:
    # Note that this used to order the body based on groupings, but now that occurs in the
    # renderer itself.
    return body.__class__(copy_data(body.body))
