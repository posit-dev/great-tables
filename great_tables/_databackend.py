from __future__ import annotations

import importlib
from abc import ABCMeta
from types import ModuleType


def _load_class(mod_name: str, cls_name: str) -> type:
    # follow the path of mods if it is one
    listified_mod_path: list[str] = cls_name.split(".")

    sub_module_path: list[str] = [mod_name] + listified_mod_path[:-1]

    cur_mod: ModuleType = importlib.import_module(sub_module_path[0])
    _cls_name: str = listified_mod_path[-1]

    if len(sub_module_path) == 1:  # no submodules, just the module itself
        return getattr(cur_mod, _cls_name)

    # now, we must increment the module paths to import the final class
    remaining_mods: list[str] = sub_module_path[1:]
    for mod in remaining_mods:
        cur_mod: ModuleType = getattr(cur_mod, mod)

    return getattr(cur_mod, _cls_name)


class _AbstractBackendMeta(ABCMeta):
    def register_backend(cls, mod_name: str, cls_name: str):
        cls._backends.append((mod_name, cls_name))
        cls._abc_caches_clear()


class AbstractBackend(metaclass=_AbstractBackendMeta):
    @classmethod
    def __init_subclass__(cls):
        if not hasattr(cls, "_backends"):
            cls._backends = []

    @classmethod
    def __subclasshook__(cls, subclass):
        for mod_name, cls_name in cls._backends:
            try:
                parent_candidate = _load_class(mod_name, cls_name)
                if issubclass(subclass, parent_candidate):
                    return True
            except (ModuleNotFoundError, ImportError):
                continue

        return NotImplemented
