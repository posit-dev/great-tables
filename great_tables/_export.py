from __future__ import annotations
from ._gt_data import GTData
from ._utils import _try_import
from typing import Optional, TYPE_CHECKING

import tempfile

if TYPE_CHECKING:
    # Note that as_raw_html uses methods on the GT class, not just data
    from .gt import GT


def as_raw_html(
    self: GT,
    make_page: bool = False,
    all_important: bool = False,
) -> str:
    """
    Get the HTML content of a GT object.

    Get the HTML content from a GT object as a string. This function is useful for obtaining the
    HTML content of a GT object for use in other contexts.

    Parameters
    ----------
    gt
        A GT object.

    Returns
    -------
    str
        An HTML fragment containing a table.
    """
    built_table = self._build_data(context="html")

    html_table = built_table._render_as_html(
        make_page=make_page,
        all_important=all_important,
    )

    return html_table


def save(
    self: GTData,
    file: str,
    selector: str = "table",
    scale: float = 1.0,
    expand: int = 5,
    window_size: tuple[int, int] = (6000, 6000),
) -> None:
    """
    Save a table as an image file or a PDF document.

    The `save()` method makes it easy to save a table object as an image file. The function produces
    a high-resolution image file or PDF of the table. The output file is create by first taking a
    screenshot of the table using a headless Chrome browser. Then, the screenshot is then cropped to
    only include the table element. Finally, and the resulting image is saved to the specified file
    path in the format specified (via the file extension).

    Parameters
    ----------
    file
        The name of the file to save the image to. Accepts names ending with .png, .bmp, and other
        image extensions. Also accepts the extension .pdf.
    selector
        The HTML element selector to use to select the table. By default, this is set to "table",
        which selects the first table element in the HTML content.
    scale
        The scaling factor that will be used when generating the image. By default, this is set to a
        value of `1.0`. Lowering this will result in a smaller image, whereas increasing it will
        result in a much higher-resolution image. This can be considered a quality setting, yet it
        also affects the file size. The default value of `1.0` is a good balance between file size
        and quality.
    expand
        The number of pixels to expand the screenshot by. By default, this is set to 5. This can be
        increased to capture more of the surrounding area, or decreased to capture less.
    window_size
        The size of the window to use when taking the screenshot. This is a tuple of two integers,
        representing the width and height of the window. By default, this is set to `(6000, 6000)`,
        a large size that should be sufficient for most tables. If the table is larger than this
        (and this will be obvious once inspecting the image file) you can increase the appropriate
        values of the tuple. If the table is very small, then a reduction in these these values will
        result in a speed gain during image capture. Please note that the window size is *not* the
        same as the final image size. The table will be captured at the same size as it is displayed
        in the headless browser, and the window size is used to ensure that the entire table is
        visible in the screen capture before the cropping process occurs.

    Returns
    -------
    None
        This function does not return anything; it simply saves the image to the specified file
        path.

    Details
    -------
    We create the output file based on the HTML version of the table.

    This process is facilitated by two libraries:

    - `selenium`, which is used to control the Chrome browser and take a screenshot of the table.
    - `PIL`, which is used to crop the screenshot to only include the table element of the page.

    Both of these packages needs to be installed before attempting to save any table as an image
    file. The `selenium` package also requires the Chrome browser to be installed on the system.

    A pip-based reinstallation of **Great Tables** through the following command will install these
    required packages:

    ```bash
    pip install great_tables[extra]
    ```

    """

    # Import the required packages
    _try_import(name="selenium", pip_install_line="pip install selenium")
    _try_import(name="PIL", pip_install_line="pip install pillow")

    from selenium import webdriver
    from selenium.webdriver.common.by import By
    from PIL import Image
    from io import BytesIO
    from pathlib import Path

    Image.MAX_IMAGE_PIXELS = None

    # Get the file extension from the file name
    file_extension = file.split(".")[-1]

    # If there is no file extension, add the .png extension
    if len(file_extension) == len(file):
        file += ".png"

    # Get the HTML content from the displayed output
    html_content = as_raw_html(self=self)

    # Create a temp directory to store the HTML file
    temp_dir = tempfile.mkdtemp()

    # Set up the Chrome webdriver options
    options = webdriver.ChromeOptions()

    # Use headless mode with an extremely large window size
    options.add_argument("--headless")

    # Set the window size (x by y) to a large value to ensure the entire table
    # is visible in the screenshot; this is settable via the `window_size=` argument
    options.add_argument(f"--window-size={window_size[0]},{window_size[1]}")

    with tempfile.NamedTemporaryFile(suffix=".html", dir=temp_dir) as temp_file, webdriver.Chrome(
        options=options
    ) as chrome:

        # Write the HTML content to the temp file
        with open(temp_file.name, "w") as fp:
            fp.write(html_content)

        # Convert the scale value to a percentage string used by the
        # Chrome browser for zooming
        zoom_level = str(scale * 100) + "%"

        # Get the scaling factor by multiplying `scale` by 2
        scaling_factor = scale * 2

        # Adjust the expand value by the scaling factor
        expansion_amount = expand * scaling_factor

        # Open the HTML file in the Chrome browser
        chrome.get("file://" + temp_file.name)
        chrome.execute_script(f"document.body.style.zoom = '{zoom_level}'")

        # Get only the chosen element from the page; by default, this is the table element
        element = chrome.find_element(by=By.TAG_NAME, value=selector)

        # Get the location and size of the table element; this will be used
        # to crop the screenshot later
        location = element.location
        size = element.size

        # Get a screenshot of the entire page as a PNG image
        png = chrome.get_screenshot_as_png()

    # Open the screenshot as an image with the PIL library; since the screenshot will be large
    # (due to the large window size), we use the BytesIO class to handle the large image data
    image = Image.open(fp=BytesIO(png))

    # Crop the image to only include the table element; the scaling factor
    # of 6 is used to account for the zoom level of 300% set earlier
    left = (location["x"] * scaling_factor) - expansion_amount
    top = (location["y"] * scaling_factor) - expansion_amount
    right = ((location["x"] + size["width"]) * scaling_factor) + expansion_amount
    bottom = ((location["y"] + size["height"]) * scaling_factor) + expansion_amount

    # Save the cropped image to the output path
    image = image.crop((left, top, right, bottom))

    # Save the image to the output path in the specified format
    image.save(fp=file)
