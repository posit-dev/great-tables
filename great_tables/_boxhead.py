from __future__ import annotations

from typing import TYPE_CHECKING

from ._locations import resolve_cols_c
from ._utils import _assert_list_is_subset
from ._tbl_data import SelectExpr
from ._text import BaseText

if TYPE_CHECKING:
    from ._types import GTSelf


def cols_label(
    self: GTSelf, cases: dict[str, str | BaseText] | None = None, **kwargs: str | BaseText
) -> GTSelf:
    """
    Relabel one or more columns.

    There are three important pieces to labelling:

    * Each argument has the form: {name in data} = {new label}.
    * Multiple columns may be given the same label.
    * Labels may use curly braces to apply special formatting, called unit notation.
      For example, "area ({{ft^2}})" would appear as "area (ftÂ²)".

    See [`define_units()`](`great_tables.define_units`) for details on unit notation.

    Parameters
    ----------
    cases
        A dictionary where the keys are column names and the values are the labels. Labels may use
        [`md()`](`great_tables.md`) or [`html()`](`great_tables.html`) helpers for formatting.

    **kwargs
        Keyword arguments to specify column labels. Each keyword corresponds to a column name, with
        its value indicating the new label.

    Returns
    -------
    GT
        The GT object is returned. This is the same object that the method is called on so that we
        can facilitate method chaining.

    Notes
    -----
    GT always selects columns using their name in the underlying data. This means that a column's
    label is purely for final presentation.

    Examples
    --------

    The example below relabels columns from the `countrypops` data to start with uppercase.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "Uganda"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Country Name",
            year="Year",
            population="Population"
        )
    )
    ```

    Note that we supplied the name of the column as the key, and the new label as the value.

    We can also use Markdown formatting for the column labels. In this example, we'll use
    `md("*Population*")` to make the label italicized.

    ```{python}
    from great_tables import GT, md
    from great_tables.data import countrypops

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Name",
            year="Year",
            population=md("*Population*")
        )
    )
    ```

    We can also use unit notation to format the column labels. In this example, we'll use
    `{{cm^3 molecules^-1 s^-1}}` for part of the label for the `OH_k298` column.

    ```{python}
    from great_tables import GT
    from great_tables.data import reactions
    import polars as pl

    reactions_mini = (
        pl.from_pandas(reactions)
        .filter(pl.col("cmpd_type") == "mercaptan")
        .select(["cmpd_name", "OH_k298"])
    )

    (
        GT(reactions_mini)
        .fmt_scientific("OH_k298")
        .sub_missing()
        .cols_label(
            cmpd_name="Compound Name",
            OH_k298="OH, {{cm^3 molecules^-1 s^-1}}",
        )
    )
    ```
    """
    from great_tables._helpers import UnitStr

    cases = cases if cases is not None else {}
    new_cases = cases | kwargs

    # If nothing is provided, return `data` unchanged
    if len(new_cases) == 0:
        return self

    # Get the full list of column names for the data
    column_names = self._boxhead._get_columns()
    mod_columns = list(new_cases.keys())

    # Stop function if any of the column names specified are not in `cols_labels`
    # msg: "All column names provided must exist in the input `.data` table."
    _assert_list_is_subset(mod_columns, set_list=column_names)

    # Handle units syntax in labels (e.g., "Density ({{ppl / mi^2}})")
    new_kwargs: dict[str, UnitStr | str | BaseText] = {}

    for k, v in new_cases.items():
        if isinstance(v, str):
            unitstr_v = UnitStr.from_str(v)

            if len(unitstr_v.units_str) == 1 and isinstance(unitstr_v.units_str[0], str):
                new_kwargs[k] = unitstr_v.units_str[0]
            else:
                new_kwargs[k] = unitstr_v

        elif isinstance(v, BaseText):
            new_kwargs[k] = v

        else:
            raise ValueError(
                "Column labels must be strings or BaseText objects. Use `md()` or `html()` for formatting."
            )

    boxhead = self._boxhead._set_column_labels(new_kwargs)

    return self._replace(_boxhead=boxhead)


def cols_align(self: GTSelf, align: str = "left", columns: SelectExpr = None) -> GTSelf:
    """
    Set the alignment of one or more columns.

    The `cols_align()` method sets the alignment of one or more columns. The `align` argument
    can be set to one of `"left"`, `"center"`, or `"right"` and the `columns` argument can be
    used to specify which columns to apply the alignment to. If `columns` is not specified, the
    alignment is applied to all columns.

    Parameters
    ----------
    align
        The alignment to apply. Must be one of `"left"`, `"center"`, or `"right"`.
    columns
        The columns to target. Can either be a single column name or a series of column names
        provided in a list. If `None`, the alignment is applied to all columns.

    Returns
    -------
    GT
        The GT object is returned. This is the same object that the method is called on so that we
        can facilitate method chaining.

    Examples
    --------
    Let's use the `countrypops` to create a small table. We can change the alignment of the
    `population` column with `cols_align()`. In this example, the column label and body cells of
    `population` will be aligned to the left.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "San Marino"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini, rowname_col="year", groupname_col="country_name")
        .cols_align(align="left", columns="population")
    )
    ```

    """

    # Throw if `align` is not one of the three allowed values
    if align not in ["left", "center", "right"]:
        raise ValueError("Align must be one of 'left', 'center', or 'right'.")

    # Get the full list of column names for the data
    column_names = self._boxhead._get_columns()

    # Upgrade `columns` to a list if `columns` is a string and not None
    if isinstance(columns, str):
        columns = [columns]
        _assert_list_is_subset(columns, set_list=column_names)
    elif columns is None:
        columns = column_names

    sel_cols = resolve_cols_c(data=self, expr=columns)

    # Set the alignment for each column
    return self._replace(_boxhead=self._boxhead._set_column_aligns(sel_cols, align=align))
