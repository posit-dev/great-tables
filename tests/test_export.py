import pytest
import requests
import sys
import tempfile
import time

from great_tables import GT, exibble, md
from great_tables._export import _infer_render_target, _create_temp_file_server
from pathlib import Path

from IPython.terminal.interactiveshell import TerminalInteractiveShell, InteractiveShell
from ipykernel.zmqshell import ZMQInteractiveShell


@pytest.fixture
def gt_tbl():
    gt_tbl = (
        GT(
            exibble[["num", "char", "currency", "row", "group"]],
            rowname_col="row",
            groupname_col="group",
            id="test_table",
        )
        .tab_header(
            title=md("Data listing from **exibble**"),
            subtitle=md("`exibble` is a **Great Tables** dataset."),
        )
        .fmt_number(columns="num")
        .fmt_currency(columns="currency")
        .tab_source_note(source_note="This is only a subset of the dataset.")
    )

    return gt_tbl


def test_html_string_generated(gt_tbl: GT, snapshot: str):
    assert snapshot == gt_tbl.as_raw_html()


@pytest.mark.skipif(sys.platform == "win32", reason="chrome might not be installed.")
@pytest.mark.extra
def test_save_image_file(gt_tbl: GT, tmp_path):

    f_path = tmp_path / "test_image.png"
    gt_tbl.save(file=str(f_path))

    time.sleep(0.05)
    assert f_path.exists()


def test_save_non_png(gt_tbl: GT, tmp_path):
    f_path = tmp_path / "test_image.pdf"
    gt_tbl.save(file=str(f_path))


def test_save_custom_webdriver(gt_tbl: GT, tmp_path):
    from selenium import webdriver

    f_path = tmp_path / "test_image.png"
    options = webdriver.ChromeOptions()
    options.add_argument("--headless=new")
    with webdriver.Chrome(options) as wd:
        gt_tbl.save(file=str(f_path), web_driver=wd)

    time.sleep(0.05)
    assert f_path.exists()


@pytest.mark.parametrize(
    "src, dst",
    [
        (InteractiveShell, "notebook"),
        (TerminalInteractiveShell, "browser"),
        (ZMQInteractiveShell, "notebook"),
        (None, "browser"),
    ],
)
def test_infer_render_target(src, dst):
    shell = src() if src is not None else src
    assert _infer_render_target(shell) == dst


def test_create_temp_file_server():
    from threading import Thread

    with tempfile.TemporaryDirectory() as tmp_dir:
        p_file = Path(tmp_dir, "index.html")
        p_file.write_text("abc")
        server = _create_temp_file_server(p_file)
        thread = Thread(target=server.handle_request)
        thread.start()

        time.sleep(0.3)
        r = requests.get(f"http://127.0.0.1:{server.server_port}/{p_file.name}")
        r.raise_for_status()
        r.content.decode() == "abc"

        thread.join()
