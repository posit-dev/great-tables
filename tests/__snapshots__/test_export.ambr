# serializer version: 1
# name: test_html_string_generated
  '''
  <div id="test_table" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
  <style>
  #test_table table {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
            -webkit-font-smoothing: antialiased;
            -moz-osx-font-smoothing: grayscale;
          }
  
  #test_table thead, tbody, tfoot, tr, td, th { border-style: none; }
   tr { background-color: transparent; }
  #test_table p { margin: 0; padding: 0; }
   #test_table .gt_table { display: table; border-collapse: collapse; line-height: normal; margin-left: auto; margin-right: auto; color: #333333; font-size: 16px; font-weight: normal; font-style: normal; background-color: #FFFFFF; width: auto; border-top-style: solid; border-top-width: 2px; border-top-color: #A8A8A8; border-right-style: none; border-right-width: 2px; border-right-color: #D3D3D3; border-bottom-style: solid; border-bottom-width: 2px; border-bottom-color: #A8A8A8; border-left-style: none; border-left-width: 2px; border-left-color: #D3D3D3; }
   #test_table .gt_caption { padding-top: 4px; padding-bottom: 4px; }
   #test_table .gt_title { color: #333333; font-size: 125%; font-weight: initial; padding-top: 4px; padding-bottom: 4px; padding-left: 5px; padding-right: 5px; border-bottom-color: #FFFFFF; border-bottom-width: 0; }
   #test_table .gt_subtitle { color: #333333; font-size: 85%; font-weight: initial; padding-top: 3px; padding-bottom: 5px; padding-left: 5px; padding-right: 5px; border-top-color: #FFFFFF; border-top-width: 0; }
   #test_table .gt_heading { background-color: #FFFFFF; text-align: center; border-bottom-color: #FFFFFF; border-left-style: none; border-left-width: 1px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 1px; border-right-color: #D3D3D3; }
   #test_table .gt_bottom_border { border-bottom-style: solid; border-bottom-width: 2px; border-bottom-color: #D3D3D3; }
   #test_table .gt_col_headings { border-top-style: solid; border-top-width: 2px; border-top-color: #D3D3D3; border-bottom-style: solid; border-bottom-width: 2px; border-bottom-color: #D3D3D3; border-left-style: none; border-left-width: 1px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 1px; border-right-color: #D3D3D3; }
   #test_table .gt_col_heading { color: #333333; background-color: #FFFFFF; font-size: 100%; font-weight: normal; text-transform: inherit; border-left-style: none; border-left-width: 1px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 1px; border-right-color: #D3D3D3; vertical-align: bottom; padding-top: 5px; padding-bottom: 5px; padding-left: 5px; padding-right: 5px; overflow-x: hidden; }
   #test_table .gt_column_spanner_outer { color: #333333; background-color: #FFFFFF; font-size: 100%; font-weight: normal; text-transform: inherit; padding-top: 0; padding-bottom: 0; padding-left: 4px; padding-right: 4px; }
   #test_table .gt_column_spanner_outer:first-child { padding-left: 0; }
   #test_table .gt_column_spanner_outer:last-child { padding-right: 0; }
   #test_table .gt_column_spanner { border-bottom-style: solid; border-bottom-width: 2px; border-bottom-color: #D3D3D3; vertical-align: bottom; padding-top: 5px; padding-bottom: 5px; overflow-x: hidden; display: inline-block; width: 100%; }
   #test_table .gt_spanner_row { border-bottom-style: hidden; }
   #test_table .gt_group_heading { padding-top: 8px; padding-bottom: 8px; padding-left: 5px; padding-right: 5px; color: #333333; background-color: #FFFFFF; font-size: 100%; font-weight: initial; text-transform: inherit; border-top-style: solid; border-top-width: 2px; border-top-color: #D3D3D3; border-bottom-style: solid; border-bottom-width: 2px; border-bottom-color: #D3D3D3; border-left-style: none; border-left-width: 1px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 1px; border-right-color: #D3D3D3; vertical-align: middle; text-align: left; }
   #test_table .gt_empty_group_heading { padding: 0.5px; color: #333333; background-color: #FFFFFF; font-size: 100%; font-weight: initial; border-top-style: solid; border-top-width: 2px; border-top-color: #D3D3D3; border-bottom-style: solid; border-bottom-width: 2px; border-bottom-color: #D3D3D3; vertical-align: middle; }
   #test_table .gt_from_md> :first-child { margin-top: 0; }
   #test_table .gt_from_md> :last-child { margin-bottom: 0; }
   #test_table .gt_row { padding-top: 8px; padding-bottom: 8px; padding-left: 5px; padding-right: 5px; margin: 10px; border-top-style: solid; border-top-width: 1px; border-top-color: #D3D3D3; border-left-style: none; border-left-width: 1px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 1px; border-right-color: #D3D3D3; vertical-align: middle; overflow-x: hidden; }
   #test_table .gt_stub { color: #333333; background-color: #FFFFFF; font-size: 100%; font-weight: initial; text-transform: inherit; border-right-style: solid; border-right-width: 2px; border-right-color: #D3D3D3; padding-left: 5px; padding-right: 5px; }
   #test_table .gt_stub_row_group { color: #333333; background-color: #FFFFFF; font-size: 100%; font-weight: initial; text-transform: inherit; border-right-style: solid; border-right-width: 2px; border-right-color: #D3D3D3; padding-left: 5px; padding-right: 5px; vertical-align: top; }
   #test_table .gt_row_group_first td { border-top-width: 2px; }
   #test_table .gt_row_group_first th { border-top-width: 2px; }
   #test_table .gt_striped { background-color: rgba(128,128,128,0.05); }
   #test_table .gt_table_body { border-top-style: solid; border-top-width: 2px; border-top-color: #D3D3D3; border-bottom-style: solid; border-bottom-width: 2px; border-bottom-color: #D3D3D3; }
   #test_table .gt_sourcenotes { color: #333333; background-color: #FFFFFF; border-bottom-style: none; border-bottom-width: 2px; border-bottom-color: #D3D3D3; border-left-style: none; border-left-width: 2px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 2px; border-right-color: #D3D3D3; }
   #test_table .gt_sourcenote { font-size: 90%; padding-top: 4px; padding-bottom: 4px; padding-left: 5px; padding-right: 5px; text-align: left; }
   #test_table .gt_left { text-align: left; }
   #test_table .gt_center { text-align: center; }
   #test_table .gt_right { text-align: right; font-variant-numeric: tabular-nums; }
   #test_table .gt_font_normal { font-weight: normal; }
   #test_table .gt_font_bold { font-weight: bold; }
   #test_table .gt_font_italic { font-style: italic; }
   #test_table .gt_super { font-size: 65%; }
   #test_table .gt_footnote_marks { font-size: 75%; vertical-align: 0.4em; position: initial; }
   #test_table .gt_asterisk { font-size: 100%; vertical-align: 0; }
   
  </style>
  <table class="gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
  <thead>
  
    <tr class="gt_heading">
      <td colspan="4" class="gt_heading gt_title gt_font_normal">Data listing from <strong>exibble</strong></td>
    </tr>
    <tr class="gt_heading">
      <td colspan="4" class="gt_heading gt_subtitle gt_font_normal gt_bottom_border"><code>exibble</code> is a <strong>Great Tables</strong> dataset.</td>
    </tr>
  <tr class="gt_col_headings">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id=""></th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="num">num</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col" id="char">char</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="currency">currency</th>
  </tr>
  </thead>
  <tbody class="gt_table_body">
    <tr class="gt_group_heading_row">
      <th class="gt_group_heading" colspan="4">grp_a</th>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">row_1</th>
      <td class="gt_row gt_right">0.11</td>
      <td class="gt_row gt_left">apricot</td>
      <td class="gt_row gt_right">$49.95</td>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">row_2</th>
      <td class="gt_row gt_right">2.22</td>
      <td class="gt_row gt_left">banana</td>
      <td class="gt_row gt_right">$17.95</td>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">row_3</th>
      <td class="gt_row gt_right">33.33</td>
      <td class="gt_row gt_left">coconut</td>
      <td class="gt_row gt_right">$1.39</td>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">row_4</th>
      <td class="gt_row gt_right">444.40</td>
      <td class="gt_row gt_left">durian</td>
      <td class="gt_row gt_right">$65,100.00</td>
    </tr>
    <tr class="gt_group_heading_row">
      <th class="gt_group_heading" colspan="4">grp_b</th>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">row_5</th>
      <td class="gt_row gt_right">5,550.00</td>
      <td class="gt_row gt_left"><NA></td>
      <td class="gt_row gt_right">$1,325.81</td>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">row_6</th>
      <td class="gt_row gt_right"><NA></td>
      <td class="gt_row gt_left">fig</td>
      <td class="gt_row gt_right">$13.26</td>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">row_7</th>
      <td class="gt_row gt_right">777,000.00</td>
      <td class="gt_row gt_left">grapefruit</td>
      <td class="gt_row gt_right"><NA></td>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">row_8</th>
      <td class="gt_row gt_right">8,880,000.00</td>
      <td class="gt_row gt_left">honeydew</td>
      <td class="gt_row gt_right">$0.44</td>
    </tr>
  </tbody>
    <tfoot class="gt_sourcenotes">
    
    <tr>
      <td class="gt_sourcenote" colspan="4">This is only a subset of the dataset.</td>
    </tr>
  
  </tfoot>
  
  </table>
  
  </div>
          
  '''
# ---
